### PR TITLE
Bugfix: GeneratorEnqueuer should not change the generated batches' order when workers > 1

### PR DESCRIPTION
**Pull request by GPhilo**
This is a proposed fix for the erroneous behaviour described in #6745.
I ran tests on Ubuntu 16.04 with python 3.5 and 2.7, all tests pass.
I also ran tests on Windows with python 3.5 (Tensorflow currently does not support python 2.7 on Windows) and all tests except those using multiprocessing pass.
The tests that fail are affected by the error mentioned in #5071.

**Description of the bug fix**: The idea is to exploit `multiprocessing`'s `Pool.apply_asynch` which does not block and immediately returns a `multiprocessing.pool.ApplyResult` object that is put in the queue.
The generator's batches order is kept by using a lock that guarantees that no thread can get a new batch until the last batch's `ApplyResult` object has been put in the queue.
Note that to keep the `GeneratorEnqueuer`'s external interface the same, I had to define two "private" classes that redefine the queues' `get()` method to actually return `ApplyResult.get()` (which may block if the result is not ready yet).

**NOTE**: Incidentally, fixing the original bug also led to the discovery and subsequent fix of a second bug.

**TL;DR**: The second bug would cause a `GeneratorEnqueuer` instance to stall potentially causing a deadlock. Now this cannot happen anymore.

**Second bug explanation:**
If:
1) the queue is full
2) a thread tries to `put()` and blocks until the queue has space again
3) the `GeneratorEnqueuer` instance calls `stop()`.

the thread calling `stop()` would block waiting for the thread waiting on `put()` to unblock. This would happen only if some thread calls `pop()` on the queue. This might lead to deadlock if the only thread popping from the queue is the thread calling `stop()`.

This behaviour is now prevented by the lock before the test on the queue length, which ensures no thread will ever try to `put()` when the queue is full.

### CODE REVIEW

**Comment by GPhilo** [comment]
I missed some tests failing when multiprocessing is used, due to a misconfiguration in my test environment setup. I'll close the PR until I solve the problem.
